### ISSUE TITLE
Reinstalls all the parsers every time

### ISSUE THREAD

**Issue by moncheeta** [nvim-treesitter/nvim-treesitter#3605]
### Describe the bug

Every time I open neovim, all the parsers are reinstalled.

### To Reproduce

1. Open neovim

### Expected behavior

The parsers are supposed to install once.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.6 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v18.9.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `gcc` executable found. Selected from { "gcc" }
    Version: [?7l[1m[34m  \\  \\ //
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.7.2
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by nixbld

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/nix/store/1s81xj0cbs0afh58r2k1x099b7h98rhh-neovim-unwrapped-0.7.2/share/nvim
"

Run :checkhealth for more info
```

### Additional context

I'm using NixOS.

Part of my init.lua: 
``
require ("nvim-treesitter.install").compilers = { "gcc" }
require("orgmode").setup_ts_grammar()
require("nvim-treesitter.configs").setup {
  highlight = {
    enable = true,
    additional_vim_regex_highlighting = { "org" },
  },
  ensure_installed = "all",
  parser_install_dir = "~/nvim",
}
``

I have to create a custom directory for the parsers since the package directory is read only in nix.

**Comment by bondzula** [comment]
I'm having the same issue, also on NixOS, set my install directory to `~/.local/share/nvim`. Seems like the issue is that treesitter does not detect installed parsers at all, like in the checkhealth screenshot, its not showing any to me, even tho i have them in the folder, and that's why its reinstalling every time. Also highlighting is not working for the same reason i guess

**Comment by dj95** [comment]
I experienced the same issue. In my case, I appended the custom path to the runtime path after calling tree-sitter's setup function. This caused a confusion, since the path is not available while checking for the installation. 

Appending the runtime path before calling the setup function resolved the issue.

```lua
vim.opt.runtimepath:append("$HOME/.local/share/treesitter")

-- initialize and configure treesitter syntax highlightning
treesitter.setup {
  parser_install_dir = "$HOME/.local/share/treesitter",
-- ...
}
```

**Comment by theHamsta** [comment]
We mentioned in the Readme that the runtime path needs to be updated. If it needs to be done before the setup call to support`ensure_installed`. It should be updated.

Would one you Nix users be interested to create a new Flake for the parsers (individually or the bundle) or all of nvim-treesitter to provide a better experience for Nix users?

**Comment by dj95** [comment]
@theHamsta I really appreciate the idea and it would be great to have the versions of tree-sitter parsers pinned as nix packages. But I'm not sure if the maintenance overhead would be feasible since a new flake would not resolve the issue.

The main problem on NixOS is, that nvim-treesitter installs parsers in the plugin directory per default. Since the nix store is read-only, this will result in an error. Therefore setting the `parser_install_dir` is required and imho a good way to mitigate the issue. 

I'm not aware of a method, how we could introduce this config change with a flake, since I only have very basic knowledge in building a flake.

Maybe we could mention a hint in the installation section of the README, which then links to the advanced configuration for setting the `parser_install_dir`. What is your opinion on that? Should I go on with this solution and create a pull request?

**Comment by bondzula** [comment]
Setting the runtime path before calling the setup method worked for me, thanks!

**Comment by Moncheeta** [comment]
I tried appending the runtime path but I get:
```
line   26:
E5113: Error while calling lua chunk: ...eta/Config/NixOS/modules/programs/editor/neovim/init.lua:241: a
ttempt to call method 'append' (a nil value)
stack traceback:
        ...eta/Config/NixOS/modules/programs/editor/neovim/init.lua:241: in main chunk
```

**Comment by dvzrv** [comment]
I have a setup in which I append the custom `parser_install_dir`  to `runtimepath` *before* adding the setup for treesitter.

However, nvim-treesitter tries to install all parsers either way, leading to a lot of failed attempts (for different reasons) and longer startup times.

`:checkhealth nvim-treesitter` shows them as being installed though (and the `.so` files are compiled already).

**Comment by lewis6991** [comment]
Is this still an issue?

**Comment by rodhash** [comment]
> We mentioned in the Readme that the runtime path needs to be updated. If it needs to be done before the setup call to support`ensure_installed`. It should be updated.
> 
> Would one you Nix users be interested to create a new Flake for the parsers (individually or the bundle) or all of nvim-treesitter to provide a better experience for Nix users?

Are we supposed to update the runtime path only for custom paths?

After migrating to Mac M1 I started facing this issue which never had on my N Linux machines.. so I tried setting up the runtime path anyway and pointed to the path already in use by treesitter .. apparently fixed for me

**Comment by javiertury** [comment]
I am running circles between two problems. When I use @dj95 solution https://github.com/nvim-treesitter/nvim-treesitter/issues/3605#issuecomment-1299954399, there are no more reinstalls, however nvim has problems related to `query error: invalid node type at position`.

The readme https://github.com/nvim-treesitter/nvim-treesitter#i-get-query-error-invalid-node-type-at-position says that there should be using only one runtime directory. The output from  `:echo nvim_get_runtime_file('parser', v:true)` instead shows two directories, the first corresponds to the installation of `nvim-treesitter` by `lazy` and the second corresponds to the additional runtimepath from @dj95 solution.

Has anybody managed to solve both problems at the same time?

**Comment by dj95** [comment]
@javiertury I had a similar issue when setting up tree-sitter with nix and finding this workaround. In my case the problem was, as in your example, another directory that contained older parsers. These older parsers were loaded in favour of the one's in the configured `parser_install_dir`.

Could you check your other runtime paths for older parsers and remove them?

**Comment by javiertury** [comment]
I tried to remove from `runtimepath` the route to the installation of `nvim-treesitter` inside  `lazy`. Now the errors reported by `:checkhealth` are gone and `:echo nvim_get_runtime_file('parser', v:true) ` reports only one directory, but opening files still shows many `query error: invalid node type at position`.

```
local parser_install_dir = vim.fn.stdpath("data") .. "/treesitter"

vim.opt.runtimepath:append(parser_install_dir)

for k, v in ipairs(vim.opt.runtimepath:get()) do
  if (string.match(v, "lazy/nvim%-treesitter")) then
    vim.opt.runtimepath:remove(v)
  end
end

-- TreeSitter setup
require("nvim-treesitter.configs").setup {
  parser_install_dir = parser_install_dir,
  ...
```

**Comment by clason** [comment]
Duplicate of #3092
